Implement NATS JetStream Destination

alyakimenko opened the pull request:
### Description

Implement NATS JetStream Destination.

### Quick checks:

- [x] I have followed the [Code Guidelines](https://github.com/ConduitIO/conduit/blob/main/docs/code_guidelines.md).
- [x] There is no other [pull request](https://github.com/ConduitIO/conduit-connector-s3/pulls) for the same update/change.
- [x] I have written unit tests.
- [x] I have made sure that the PR is of reasonable size and can be easily reviewed.

hariso commented on destination/destination_integration_test.go:
```diff
@@ -0,0 +1,223 @@
+	})
+	is.NoErr(err)
+
+	t.Run("success, jetstream", func(t *testing.T) {
```
Personally, whenever I see multiple `t.Run()` calls, I think if it makes sense having a table driven test, or just having separate test methods.

alyakimenko replied:
I've split it into separate test functions in the https://github.com/conduitio-labs/conduit-connector-nats-jetstream/pull/10/commits/5d190f808b484d02f22e7294ea29801227c75c1b

hariso commented on destination/destination_integration_test.go:
```diff
@@ -0,0 +1,223 @@
+		is.NoErr(err)
+	})
+
+	t.Run("fail, jetstream try async write, 1 message, batch is 1", func(t *testing.T) {
```
Why is an error expected here?

alyakimenko commented on destination/destination.go:
```diff
@@ -0,0 +1,117 @@
+	return nil
+}
+
+// Write writes a record into a Destination.
+func (d *Destination) Write(ctx context.Context, record sdk.Record) error {
+	return d.writer.Write(ctx, record)
+}
+
+// WriteAsync asynchronously writes a record into a Destination.
+// JetStream supports it when the batchSize is greater than 1, otherwise the method will fallback to Write.
+// When a batch is full the method calls Flush.
+func (d *Destination) WriteAsync(ctx context.Context, record sdk.Record, ackFunc sdk.AckFunc) error {
```
Yeah, they are. If the `batchSize` is greater than 1 then we'll use the `Write`, otherwise, we'll use `WriteAsync`.
We have the following checks within the Writer's methods:
`Write` method:
```go
if w.batchSize > 1 {
		return sdk.ErrUnimplemented
}
```
`WriteAsync` method:
```go
if w.batchSize == 1 {
		return false, sdk.ErrUnimplemented
}
```

alyakimenko commented on destination/destination_integration_test.go:
```diff
@@ -0,0 +1,223 @@
+	})
+	is.NoErr(err)
+
+	t.Run("success, jetstream sync write, 1 message, batch size is 1", func(t *testing.T) {
```
Fixed in the https://github.com/conduitio-labs/conduit-connector-nats-jetstream/pull/10/commits/5d190f808b484d02f22e7294ea29801227c75c1b

alyakimenko commented on acceptance_test.go:
```diff
@@ -0,0 +1,83 @@
+				DestinationConfig: cfg,
+				BeforeTest:        beforeTest(t, cfg),
+				GoleakOptions: []goleak.Option{
+					// nats.go spawns a separate goroutine to process flush requests
```
The goroutine should stop when we call `*nats.Conn.Stop` method, but it seems that there's something broken in the library

alyakimenko replied:
I've added a note about it to a comment

alyakimenko commented on README.md:
```diff
@@ -46,16 +46,42 @@ The config passed to Configure can contain the following fields.
 | `urls`                     | A list of connection URLs joined by comma. Must be a valid URLs.<br />Examples:<br />`nats://127.0.0.1:1222`<br />`nats://127.0.0.1:1222,nats://127.0.0.1:1223`<br />`nats://myname:password@127.0.0.1:4222`<br />`nats://mytoken@127.0.0.1:4222`                                                                                                                                ...
 | `streamName`               | A stream name. It can contains alphanumeric characters and max length is 32 characters.                                                                                                                                                                                                                                                                                          ...
-| `subject`                  | A name of a subject from which or to which the connector should read write.                                                                                                                                                                                                                                                                                                      ...
+| `subject`                  | A name of a subject from which or to which the connector should read write. It's possible to specify a subject that is different from the one belonging to the stream you specified, the connector in this case will handle messages properly.                                                                                                                                   ...
```
Sorry, I guess I misrepresented this behavior. I meant that it is possible to specify a name of a subject that belongs to a stream, but not to the stream you specified, and the connector will still read messages from the subject. This is the way how the nats.go library works by default.

Another option here is to retrieve `StreamInfo` by a stream name and then check if its `subjects` slice contains the subject you specified in the config and if it is not - raise an error.

What do you think? Should I add the check?

alyakimenko replied:
I've removed the `stream` parameter and added a new one, `deliverSubject`, which is `<durable>.conduit` by default. The `stream` field was only used in constructing the deliver subject during the channel subscription, so yeah, having the `deliverSubject` instead of the `stream` makes much more sense.

alyakimenko commented:
@hariso In addition to the fixes above, I've also updated the SDK to the latest version. I've simplified the `Destination`, removed the `WriteAsync` and `Flush` methods, and made message sending synchronous.

hariso commented on destination/config.go:
```diff
@@ -0,0 +1,93 @@
+}
+
+// Parse maps the incoming map to the Config and validates it.
+func Parse(cfg map[string]string) (Config, error) {
```
WDYT about using JSON unmarshaling to try simplifying the parsing logic?

hariso replied:
I was thinking of marshalling the map into JSON bytes, and then unmarshalling into a struct. It's an additional step, but  I think it would save all the lines where parse fields one by one (parse int, parse duration etc.)

alyakimenko replied:
But if we have, for example:
```go
cfg := map[string]string{
    "retryWait":     "1000",
    "retryAttempts": "10",
}
```

After marshalling the map into JSON bytes, we will have the following data (converted to a string): `{"retryWait": "1000", "retryAttempts": "10"}`. 
But unmarshalling it to a struct will fail trying to unmarshal string into integer, or string into `time.Duration`, etc.

hariso replied:
Ah, good point. :+1:

hariso commented on destination/destination.go:
```diff
@@ -0,0 +1,167 @@
+
+// Write writes a record into a Destination.
+func (d *Destination) Write(ctx context.Context, records []sdk.Record) (int, error) {
+	for i, record := range records {
+		if err := d.writer.Write(ctx, record); err != nil {
+			return i, fmt.Errorf("write: %w", err)
+		}
+	}
```
Is it possible to write all the records in a single batch?

alyakimenko replied:
It is possible in asynchronous mode, but not in synchronous

hariso commented on destination/destination_integration_test.go:
```diff
@@ -0,0 +1,120 @@
+	is.NoErr(err)
+}
+
+func TestDestination_Open_Fail(t *testing.T) {
```
It would be good to explain why is this test expect to fail (either through a comment or in the method name). I guess it's because of the wrong URL?

alyakimenko replied:
Yeah, you're right, thanks, I've added a comment explaining it

hariso commented on destination/jetstream/writer.go:
```diff
@@ -0,0 +1,93 @@
+}
+
+// getPublishOptions returns a NATS publish options based on the provided WriterParams.
+func getPublishOptions(params WriterParams) []nats.PubOpt {
```
Would it make sense to have this as a method in WriterParams?

alyakimenko replied:
I like the idea, thank you! I've applied it to the `WriterParams` as well as to the `IteratorParams`.
